### PR TITLE
Add support for FreeBSD 13

### PR DESCRIPTION
**Pull request by ethanuo**
**Is your pull request related to a problem? Please describe.**
No.

**Description**
Allow permasigner to run on FreeBSD 13-STABLE (and probably -RELEASE too).  
Packaging, signing, and installing are tested and work as expected

**Notes:** 

- ~ldid needs to be compiled manually and put into \~/.local/share/permasigner/ to work, since there is no FreeBSD binary in the [ldid fork](https://github.com/itsnebulalol/ldid/releases/latest) that it downloads from.~ _in progress_
- ~[comms/usbmuxd](https://www.freshports.org/comms/usbmuxd) manually needs to be ran to open /var/run/usbmuxd (it should also pull [comms/libimobiledevice](https://www.freshports.org/comms/libimobiledevice) as a dependency which is needed as well)~ _fixed_
- ~iproxy 2222 22 also needs to be run beforehand~ _fixed_

**Checklist**

-   [x] The new code was tested.
-   [x] The code isn't overly messy.

### CODE REVIEW

**Comment by ethanuo** [comment]
I managed to get the installer to run successfully, but the if loop throws the following error:
`File "/usr/home/ethan/Compiling/permasigner/permasigner/ps_installer.py", line 108, in install_deb
    if "password".encode() in output or "Password".encode() in output:
TypeError: 'in <string>' requires string as left operand, not bytes
`
It works if I remove/edit the line entirely, but I am not sure if it will mess with things if I push it upstream

**Comment by ethanuo** [comment]
Hopefully resolved it in third-to-most recent commit and added a notice in most recent and second-to-most recent commits, I think it should be safe to merge now

**Comment by superrnovae** [comment]
Doesn't the script already relay port 2222 to 22? The tcprelay is doing that. Not sure why iproxy is needed.

Someone needs to write a workflow to compile for freebsd. Otherwise the download will fail.

usbmuxd is run automatically when a device is connected, that is the case on linux.

**Comment by ethanuo** [comment]
> usbmuxd is ran automatically when a device is connected, that is the case on linux.

@superrnovae Sadly that's not the case here: From the [libimobiledevice readme:](https://github.com/libimobiledevice/usbmuxd), usbmuxd uses systemd or udev to automatically run, but FreeBSD uses devd which works quite differently (side note - [devel/libudev-devd](https://www.freshports.org/devel/libudev-devd) is supposed to bridge that gap).  The [comms/usbmuxd](https://www.freshports.org/comms/usbmuxd) port fixes that problem due to it being a native port, but it doesn't seem to automatically run without any additional configuration - something the typical user probably won't do - so it's at least worth a mention. 

> Someone needs to write a workflow to compile for freebsd. Otherwise the download will fail.

Good idea, though I'm not sure how difficult that would be as I am not familiar with github workflows. 

>The tcprelay is already relaying port 2222 to 22. Not sure why iproxy is needed. If it fails, then you can pass --tcprelay flag to script in the form of 22:2222:127.0.0.1:/var/run/usbmuxd, the last 2 are optional.

I can't figure out why exactly either - all I know is that something in `ps_tcprelay.py` isn't doing something right (even with `--tcprelay`...!).   I spend a good few hours trying to diagnose the problem but I just got an `error 3` no matter what I did, and I couldn't find anything about it online.  I just did a sample run with iproxy running and not running and copied the debug output, maybe someone can use it  to find out what's wrong once and for all? 

[log-iproxy.txt](https://github.com/itsnebulalol/permasigner/files/9366788/log-iproxy.txt)
[log-noiproxy.txt](https://github.com/itsnebulalol/permasigner/files/9366789/log-noiproxy.txt)

I'll try to resolve conflicts and retrace my steps to work it out, but I'm puzzled.

**Comment by superrnovae** [comment]
> [log-iproxy.txt](https://github.com/itsnebulalol/permasigner/files/9366788/log-iproxy.txt)
> [log-noiproxy.txt](https://github.com/itsnebulalol/permasigner/files/9366789/log-noiproxy.txt)

judging by logs the iproxy was running in both cases.

**Comment by ethanuo** [comment]
> I'm not familiar with freebsd, is running /usr/sbin/usbmuxd enough to launch it?
yeah, though it's located in  /usr/local/sbin/usbmuxd.  It just doesn't automatically runs like people expect (I don't think I made that as clear as it should've been)

**Comment by superrnovae** [comment]
I'll debug it in the vm in the morning. In the meantime, could you please test [this artifact](https://github.com/superrnovae/ldid/suites/7864045460/artifacts/333387049)? Let me know if it doesn't work. Thank you!

**Comment by ethanuo** [comment]
The binary runs fine but it complains that ldid is malformed/outdated and tries to download it.  Apparently ps_downloader doesn't have an entry for freebsd so it freaks out and overwrites `ldid`.  Thanks for your help!

**Comment by superrnovae** [comment]
> The binary runs fine but it complains that ldid is malformed/outdated and tries to download it. Apparently `ps_downloader.py` doesn't have an entry for freebsd so it freaks out and overwrites `ldid`. Thanks for your help!

It was built in my local repo. It's was not released yet.

**Comment by superrnovae** [comment]
> > usbmuxd is ran automatically when a device is connected, that is the case on linux.
> 
> @superrnovae Sadly that's not the case here: From the [libimobiledevice readme:](https://github.com/libimobiledevice/usbmuxd) usbmuxd uses systemd or udev to automatically run, but FreeBSD uses devd (and no systemd) which works quite differently (side note - [devel/libudev-devd](https://www.freshports.org/devel/libudev-devd) is supposed to bridge that gap). The [comms/usbmuxd](https://www.freshports.org/comms/usbmuxd) port fixes that problem due to it being a native port, but it doesn't seem to automatically run without any additional configuration - something the typical user probably won't do - so it's at least worth a mention.
> 
> > Someone needs to write a workflow to compile for freebsd. Otherwise the download will fail.
> 
> Good idea, though I'm not sure how difficult that would be as I am not familiar with github workflows.
> 
> > The tcprelay is already relaying port 2222 to 22. Not sure why iproxy is needed. If it fails, then you can pass --tcprelay flag to script in the form of 22:2222:127.0.0.1:/var/run/usbmuxd, the last 2 are optional.
> 
> I can't figure out why exactly either - all I know is that something in `ps_tcprelay.py` isn't doing something right (even with `--tcprelay`...!). I spend a good few hours trying to diagnose the problem but I just got an `error 3` no matter what I did, and I couldn't find anything about it online. I just did a sample run with iproxy running and not running and copied the debug output, maybe someone can use it to find out what's wrong once and for all?
> 
> [log-iproxy.txt](https://github.com/itsnebulalol/permasigner/files/9366788/log-iproxy.txt) [log-noiproxy.txt](https://github.com/itsnebulalol/permasigner/files/9366789/log-noiproxy.txt)
> 
> I'll try to resolve conflicts and retrace my steps to work it out, but I'm puzzled.

I managed to reproduce this after uninstall openssh package from my iPad.

**Comment by ethanuo** [comment]
> > The binary runs fine but it complains that ldid is malformed/outdated and tries to download it. Apparently `ps_downloader.py` doesn't have an entry for freebsd so it freaks out and overwrites `ldid`. Thanks for your help!
> 
> It was built in my local repo. It's was not released yet.

DIfferent issue, I finally found why.  Currently, if ldid is not in path (even if it's in `data_dir`), `ldid.download()` is called (see `__main.py__` line 311) and a hash is generated from the most recent release of the fork of ldid, with a url based off what `get_arch` returns (which is either "ldid_freebsd13_x84_64" (my fork) or "nothing" (upstream)).  Since neither url actually exists, the mismatched hashes prompts the system to redownload ldid.  Somehow the url returns a 200 and ldid is overwritten, creating a file that just says 'Not Found'. 1061d00ece281135b2b315f5b422ff182a4f8388 Removed the ability for ldid to still be used anyway as a last resort  (I presume).  A manual override (possibly in the form of a flag) could bypass this, but that's probably not worth the effort

TL;DR line 311 of `__main.py__` needs an indent and the problem goes away.

> I managed to reproduce this after uninstalling openssh package from my iPad.

Which log or both?  I have openssh installed on my phone, just checked.  FWIW I have a hunch that the current implementation of tcprelay (I think) can't/isn't able to properly communicate with usbmuxd for some reason so it isn't able to establish a connection at all

**Comment by superrnovae** [comment]
> Which log or both?

the one without iproxy. 

> if ldid is not in path (even if it's in `data_dir`), `ldid.download()` is called (see `__main.py__` line 311)

Yes. That's how it works. It will fail on systems we haven't compiled ldid for yet. 

I just installed freebsd in VM. Compiled libusbmuxd, usbmuxd from ports and it connect just fine to device. It has issues sending file over scp or sftp, most likely because I'm in a vm.

**Comment by superrnovae** [comment]
I found out why tcprelay fails to connect to usbmuxd socket, I'll merge the solution later today.

**Comment by ethanuo** [comment]
> I just installed freebsd in VM. Compiled libusbmuxd, usbmuxd from ports and it connects just fine to device. I also enabled devd support and usbmuxd socket is automatically opened whenever I connect my device. But it has issues sending file over scp or sftp, whether I passthrough the usb controller or simply redirect the usb device. (I suspect it has something to do with Virtio driver)

Works for me so it's probably virtio

> Yes. That's how it works. It will fail on systems we haven't yes compiled ldid for

oh ok, I just misunderstood what it was supposed to be doing, my bad.

> Possible reason is you have more than 1 device connected. Tcprelay will always attempt to connect to first device it retrieves from the list of devices.

The only usb device connected is my phone, so that's out of the question.  Does everything work for you on FreeBSD as-is?

Might've found the root of my issue.  After playing around even more, I found that `tcpconnect.py` works whilst `ps_tcprelay.py` doesn't.   For some bizarre reason, if you replace `dsock = mux.connect(dev, self.server.server_address[1])` with  `dsock = mux.connect(dev, self.server.rport)` I get no error and the program resumes... and succeeds!  
If I understand everything correctly, `server_address[1]` is the _local_ open port (2222 for me), so by mapping that to `dsock`, tcprelay ends up trying to connect to port 2222 not 22. I was able to somewhat verify this through `pdb` (put a breakpoint at permasigner/ps_installer:45) and saw that the error was encountered while running `client.connect(` in `ps_installer.py`.

**Comment by superrnovae** [comment]
> Might've found the root of my issue.  After playing around even more, I found that `tcpconnect.py` works whilst `ps_tcprelay.py` doesn't.   ~For some bizarre reason,~ If you replace `dsock = mux.connect(dev, self.server.server_address[1])` with  `dsock = mux.connect(dev, self.server.rport)` I get no error and the program resumes... and succeeds!  This isn't wholly reliable as it sometimes freezes while copying (but not very often), but a restart of `usbmuxd` does the trick.
> 
> If I understand everything correctly, `server_address[1]` is the _local_ open port (2222 for me), so by mapping that to `dsock`, tcprelay ends up trying to connect to port 2222 not 22. I was able to somewhat verify this through `pdb` (put a breakpoint at permasigner/ps_installer:45) and saw that the error was encountered while running `client.connect(` in `ps_installer.py`.  

Btw, I've fixed that upstream.

**Comment by superrnovae** [comment]
Tcprelay issues fixed in [37de061abb5728681e8842e0cef6e7fbffccf467](https://github.com/itsnebulalol/permasigner/commit/37de061abb5728681e8842e0cef6e7fbffccf467)

**Comment by superrnovae** [comment]
As for the ldid, I'll have it compiled for both freebsd12 and freebsd 13 on x86_64 and arm64 architectures in a couple hours.

**Comment by ethanuo** [comment]
> As for the ldid, I'll have it compiled for both freebsd12 and freebsd 13 on x86_64 and arm64 architectures in a couple of hours.

Ok.  Keep in mind that FreeBSD calls `x86_64` `amd64` which may cause issues

**Comment by superrnovae** [comment]
> Ok. Keep in mind that FreeBSD calls `x86_64` `amd64` which may cause issues

I'll keep that in mind. Thanks. I'll test everything in the vm beforehand.

**Comment by ethanuo** [comment]
With 5aa881ae916db523b9afb1c9e58ebb39fceb7cdd, I think I ironed out most of the issues and the program now runs fine for me (usbmuxd can be a hit-or-miss though).  Can you run it through the VM to make sure there's nothing wrong?  You shouldn't need to mess with anything to make it work (ldid just needs to be in ~/.local/share/permasigner)

By the way, it might be a good idea to revert 411bc8f0bc545b135bfc612c6d568b2f69dfbe7b. Since that commit installed debs don't give the app execution privileges, so they just crash on startup without warning

**Comment by superrnovae** [comment]
> since that commit, installed debs don't give the app execution privileges, so they just crash on startup without warning (oh and sudo needs to be installed to make it work as well but I don't know of a workaround for that)

I know, but giving execution permission in postinst isn't an optimal solution. Some users don't have sudo package on their devices, and the install was resulting in an error for them.

**Comment by superrnovae** [comment]
I've made a PR to build ldid for FreeBSD [here](https://github.com/itsnebulalol/ldid/pull/3)

**Comment by superrnovae** [comment]
Manually merged in [67e35e4](https://github.com/itsnebulalol/permasigner/commit/67e35e470cb1be67d21ecea5ebe2d90f1db4fac3)
